esp/services/ws_config2/ws_config2Service.cpp

```diff
@@ -762,115 +760,125 @@ void Cws_config2Ex::getNodeInfo(const std::shared_ptr<SchemaItem> &pNodeSchemaIt
 }
 
 
-void Cws_config2Ex::getAttributes(const std::vector<std::shared_ptr<EnvironmentValue>> &attributes, IArrayOf<IEspAttributeType> &nodeAttributes) const
+void Cws_config2Ex::getAttributes(const std::shared_ptr<EnvironmentNode> &pEnvNode, IArrayOf<IEspAttributeType> &nodeAttributes, bool includeMissing) const
 {
-    for (auto it=attributes.begin(); it!=attributes.end(); ++it)
+    std::vector<std::shared_ptr<SchemaValue>> schemaValues;
+    pEnvNode->getSchemaItem()->getAttributes(schemaValues);
+
+    for (auto it=schemaValues.begin(); it!=schemaValues.end(); ++it)
     {
-        std::shared_ptr<EnvironmentValue> pAttr = *it;
+        const std::shared_ptr<SchemaValue> pSchemaValue = *it;
+        std::shared_ptr<EnvironmentValue> pEnvValue = pEnvNode->getAttribute(pSchemaValue->getName());
         Owned<IEspAttributeType> pAttribute = createAttributeType();
 
-        const std::shared_ptr<SchemaValue> &pSchemaValue = pAttr->getSchemaValue();
-        std::string attributeName = pAttr->getName();
-        pAttribute->setName(attributeName.c_str());
-        pAttribute->setDisplayName(pSchemaValue->getDisplayName().c_str());
-        pAttribute->setTooltip(pSchemaValue->getTooltip().c_str());
-
-        const std::shared_ptr<SchemaType> &pType = pSchemaValue->getType();
-        std::shared_ptr<SchemaTypeLimits> &pLimits = pType->getLimits();
-        pAttribute->updateType().setBaseType(pType->getBaseType().c_str());
-        pAttribute->updateType().setSubType(pType->getSubType().c_str());
-        if (pType->getLimits()->isMaxSet())
-        {
-            pAttribute->updateType().updateLimits().setMaxValid(true);
-            pAttribute->updateType().updateLimits().setMax(pType->getLimits()->getMax());
-        }
-        if (pType->getLimits()->isMinSet())
-        {
-            pAttribute->updateType().updateLimits().setMinValid(true);
-            pAttribute->updateType().updateLimits().setMin(pType->getLimits()->getMin());
-        }
-        pAttribute->setRequired(pSchemaValue->isRequired());
-        pAttribute->setReadOnly(pSchemaValue->isReadOnly());
-        pAttribute->setHidden(pSchemaValue->isHidden());
-        pAttribute->setDeprecated(pSchemaValue->isDeprecated());
-        std::string groupName = pSchemaValue->getGroup();
-        pAttribute->setGroup(groupName.empty() ? "Attributes" : groupName.c_str());
-
-        std::vector<AllowedValue> allowedValues;
-        pSchemaValue->getAllowedValues(allowedValues, pAttr->getEnvironmentNode());
-        if (!allowedValues.empty())
+        if (pEnvValue || includeMissing)
         {
-            IArrayOf<IEspChoiceType> choices;
-            for (auto valueIt=allowedValues.begin(); valueIt!=allowedValues.end(); ++valueIt)
+            std::string attributeName = pSchemaValue->getName();
```

Comment:
attributeName is only used once, why not pAttribute->setName(pSchemaValue->getName().c_str());

Reply:
There is no particular reason other than getName returns a const reference and I wanted to avoid possible compile problems down the road should attributeName be needed for another purpose. I went ahead and changed it.